### PR TITLE
Move ACCESS_TOKEN_AND_EXPIRATION_CACHE_KEY to redis

### PR DESCRIPTION
**Pull request by jeffin143**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [X] Feature
- [ ] Bug Fix
- [ ] Documentation Update

## Description
Move ACCESS_TOKEN_AND_EXPIRATION_CACHE_KEY to redis
## Related Tickets & Documents
handle #4670 

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [X] no documentation needed

### CODE REVIEW

File: app/services/streams/twitch_access_token/get.rb [hunk 1]
```diff
@@ -7,11 +7,11 @@ def self.call
       end
 
       def call
-        token, exp = Rails.cache.fetch(ACCESS_TOKEN_AND_EXPIRATION_CACHE_KEY)
+        token, exp = RedisRailsCache.fetch(ACCESS_TOKEN_AND_EXPIRATION_CACHE_KEY)
 
         if token.nil? || Time.zone.now >= exp
           token, exp = get_new_token
```
Comment (mstruve) [inline app/services/streams/twitch_access_token/get.rb]: @benhalpern question, is this expiration constantly changing and that's why we always reset it?

File: app/services/streams/twitch_access_token/get.rb [hunk 2]
```diff
@@ -7,11 +7,11 @@ def self.call
...
-          Rails.cache.write(ACCESS_TOKEN_AND_EXPIRATION_CACHE_KEY, [token, exp])
+          RedisRailsCache.write(ACCESS_TOKEN_AND_EXPIRATION_CACHE_KEY, [token, exp])
```
Comment (mstruve) [inline app/services/streams/twitch_access_token/get.rb]: I am SO SORRY this has taken a bit for us to get to! I was trying to get some additional color on the whole situation since I am still a bit new here 😃 One last thing before this is ready. This will not set the expiration like you would think according to my rough testing in a console. Let me know if that makes sense and if so we should update this method so the expiration is set correctly.
```
[38] pry(main)> RedisRailsCache.write("tok", [1, 1000])
=> "OK"
[40] pry(main)> RedisRailsCache.redis.ttl("tok")
=> 86385
[41] pry(main)> RedisRailsCache.redis.ttl("tok")
=> 86384
[42] pry(main)> RedisRailsCache.write("tok", 1, :expires_in => 1000)
=> "OK"
[43] pry(main)> RedisRailsCache.redis.ttl("tok")
=> 998
```

Comment (jeffin143) [reply app/services/streams/twitch_access_token/get.rb]: Yes, I guess you are right, How should we update it ?

Comment (rhymes) [reply app/services/streams/twitch_access_token/get.rb]: Similar to @mstruve example:

```ruby
RedisRailsCache.write(ACCESS_TOKEN_AND_EXPIRATION_CACHE_KEY, token, expires_in: exp)
```

instead of

```ruby
RedisRailsCache.write(ACCESS_TOKEN_AND_EXPIRATION_CACHE_KEY, [token, exp])
```

currently it saves an array of two values instead of the token and setting its expiration date

Comment (jeffin143) [reply app/services/streams/twitch_access_token/get.rb]: Thanks @rhymes :)

I have made apt changes @mstruve

File: app/services/streams/twitch_access_token/get.rb [hunk 3]
```diff
@@ -7,11 +7,11 @@ def self.call
       end
 
       def call
-        token, exp = Rails.cache.fetch(ACCESS_TOKEN_AND_EXPIRATION_CACHE_KEY)
+        token, exp = RedisRailsCache.fetch(ACCESS_TOKEN_AND_EXPIRATION_CACHE_KEY)
```
Comment (rhymes) [inline app/services/streams/twitch_access_token/get.rb]: btw, shouldn't this be a `.read` ?

Comment (mstruve) [reply app/services/streams/twitch_access_token/get.rb]: `fetch` will attempt to read the cache, I guess technically it could be `read` for simplicity but I think it's fine to leave it.

However, and this was a miss on my part when I suggested using the expiration as the actual expiration, we should update this cache key so that it's clear it only represents the token bc now the expiration is used as the actual expiration. We also need to calculate the time until the expiration, we can't just use a Datetime for expires. Once again, that was a review miss on my part! Sorry!

In other words, here is what I believe needs to happen to iron this out with the new approach. 

```ruby 
token = RedisRailsCache.read(ACCESS_TOKEN_CACHE_KEY)

return token unless token.nil?
token, exp = get_new_token
time_til_exp = exp - Time.now # this should give you the time until expiration in seconds which is what we need
RedisRailsCache.write(ACCESS_TOKEN_CACHE_KEY, token, expires_in: time_til_exp) # This actually returns "OK" which is why we need an additional line to return the token.
token
```

**Comment by mstruve** [comment]
Hey @jeffin143, same as the other PR, I was hoping to get this merged soon but some of the specs need updating. Would you like to fix it or do you want me to do it? I'm game either way, just let me know! 😃

**Comment by jeffin143** [comment]
> Hey @jeffin143, same as the other PR, I was hoping to get this merged soon but some of the specs need updating. Would you like to fix it or do you want me to do it? I'm game either way, just let me know! smiley

Made changes, Sorry that it took long to make changes :)

File: spec/services/streams/twitch_access_token/get_spec.rb [hunk 4]
```diff
@@ -24,7 +24,7 @@
 
     context "when there is an unexpired token in the cache" do
       it "returns the cached token" do
-        Rails.cache.write(described_class::ACCESS_TOKEN_AND_EXPIRATION_CACHE_KEY, ["FAKE_UNEXPIRED_TWITCH_TOKEN", 15.days.from_now])
+        RedisRailsCache.write(described_class::ACCESS_TOKEN_AND_EXPIRATION_CACHE_KEY, ["FAKE_UNEXPIRED_TWITCH_TOKEN", 15.days.from_now])
```
Comment (mstruve) [inline spec/services/streams/twitch_access_token/get_spec.rb]: These cache write methods are still writing the cached token the old way, as an array. We need to update them to use the new method to ensure the tests are working as expected. 

```ruby
RedisRailsCache.write(described_class::ACCESS_TOKEN_AND_EXPIRATION_CACHE_KEY, "FAKE_UNEXPIRED_TWITCH_TOKEN", :expires_in => 15.days.from_now)
```

Comment (jeffin143) [reply spec/services/streams/twitch_access_token/get_spec.rb]: Thanks, missed it :)

File: spec/services/streams/twitch_access_token/get_spec.rb [hunk 5]
```diff
@@ -33,7 +33,7 @@
 
     context "when there is an expired token in the cache" do
       it "requests a new token and caches it" do
-        RedisRailsCache.write(described_class::ACCESS_TOKEN_AND_EXPIRATION_CACHE_KEY, ["FAKE_EXPIRED_TWITCH_TOKEN", 15.days.ago])
+        RedisRailsCache.write(described_class::ACCESS_TOKEN_AND_EXPIRATION_CACHE_KEY, "FAKE_EXPIRED_TWITCH_TOKEN", :expires_in => 15.days.from_now)
```
Comment (mstruve) [inline spec/services/streams/twitch_access_token/get_spec.rb]: This is setting the expiration in the future but this test is testing what happens when a token is expired.
> "when there is an expired token in the cache"
In this case, an expired token will not exist, in other words, given no token in the cache(by removing this line) then we should return a brand new one.

Comment (jeffin143) [reply spec/services/streams/twitch_access_token/get_spec.rb]: I really didn't get that, So this line is not needed any more 

```
RedisRailsCache.write(described_class::ACCESS_TOKEN_AND_EXPIRATION_CACHE_KEY, "FAKE_EXPIRED_TWITCH_TOKEN", :expires_in => 15.days.from_now)
```

Comment (jeffin143) [reply spec/services/streams/twitch_access_token/get_spec.rb]: or will `15.days.ago work` ?

Comment (mstruve) [reply spec/services/streams/twitch_access_token/get_spec.rb]: We want to remove the write line. Creating a token in Redis with an expiration in the past doesn't really make sense bc it will just expire itself immediately and not exist. What we want to test here is what happens when a token has expired, ie does not exist in Redis. One of the easiest ways to not have it exist in Redis is to not write it in the first place.

**Comment by mstruve** [comment]
Hmmm still looks like we have a failing spec 🤔
```
  1) Streams::TwitchAccessToken::Get.access_token when there is an unexpired token in the cache returns the cached token
     Failure/Error: expect(described_class.call).to eq "FAKE_UNEXPIRED_TWITCH_TOKEN"
     
       expected: "FAKE_UNEXPIRED_TWITCH_TOKEN"
            got: "FAKE_BRAND_NEW_TWITCH_TOKEN"
     
       (compared using ==)
     # ./spec/services/streams/twitch_access_token/get_spec.rb:29:in `block (4 levels) in <top (required)>'
     # ./spec/rails_helper.rb:89:in `block (3 levels) in <top (required)>'
     # ./spec/rails_helper.rb:89:in `block (2 levels) in <top (required)>'
Finished in 12 minutes 52 seconds (files took 10.03 seconds to load)
3067 examples, 1 failure, 3 pending
Failed examples:
rspec ./spec/services/streams/twitch_access_token/get_spec.rb:26 # Streams::TwitchAccessToken::Get.access_token when there is an unexpired token in the cache returns the cached token
```
It might be helpful to run just that spec file locally and add some debugging statements like binding.pry to see what is going on bc it is unclear to me at a glance.

**Comment by mstruve** [comment]
Hey @jeffin143! As I mentioned [on your other PR](https://github.com/thepracticaldev/dev.to/pull/4820#issuecomment-563309891) we have made the switch completely to Redis so I am going to close this PR since we no longer need it. Thank you again for all the help on this project. It is contributors like yourself which make this project truly collaborative and amazing to work on! Let me know if you have any questions!!!
